contracts/0.8.25/utils/UpgradeTemplateV3.sol

```diff
@@ -0,0 +1,639 @@
+        // NB: we pass some of the addresses available in the new locator as parameters
+        //     because at the template creation time locator is not upgraded
+        _predepositGuarantee = IPredepositGuarantee(params.predepositGuarantee);
+        _burner = IBurner(params.burner);
+        _vaultFactory = IVaultFactory(params.vaultFactory);
+        _upgradeableBeacon = IUpgradeableBeacon(params.upgradeableBeacon);
+        _stakingVaultImplementation = params.stakingVaultImplementation;
+        _delegationImplementation = params.delegationImplementation;
+
+        _locator = ILidoLocator(params.locator);
+        _agent = params.agent;
+        _aragonAppLidoRepo = IAragonAppRepo(params.aragonAppLidoRepo);
+        _csmAccounting = params.csmAccounting;
+        _voting = params.voting;
+        _nodeOperatorsRegistry = params.nodeOperatorsRegistry;
+        _simpleDvt = params.simpleDvt;
+        _wstETH = params.wstETH;
+
+        _lidoImplementation = params.lidoImplementation;
+
+        _accountingOracleImplementation = params.accountingOracleImplementation;
+        _newLocatorImplementation = params.newLocatorImplementation;
+
+        _oldBurner = IBurner(_locator.burner());
+        _oldOracleReportSanityChecker = IOracleReportSanityChecker(_locator.oracleReportSanityChecker());
+        _oracleReportSanityChecker = IOracleReportSanityChecker(params.oracleReportSanityChecker);
+        _accountingOracle = IAccountingOracle(_locator.accountingOracle());
+        _elRewardsVault = _locator.elRewardsVault();
+        _stakingRouter = IStakingRouter(_locator.stakingRouter());
+        _lido = ILido(_locator.lido());
+        _validatorsExitBusOracle = _locator.validatorsExitBusOracle();
+        _withdrawalQueue = _locator.withdrawalQueue();
+    }
+
+    /// @notice Must be called after LidoLocator is upgraded
+    function startUpgrade() external {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        _assertNotExpired();
+        if (_isUpgradeFinished) revert UpgradeAlreadyFinished();
+
+        if (!ALLOW_NON_SINGLE_BLOCK_UPGRADE) {
+            if (_upgradeBlockNumber != UPGRADE_NOT_STARTED) revert UpgradeAlreadyStarted();
+            _upgradeBlockNumber = block.number;
+        }
+
+        // Save initial state for the check after burner migration
+        _initialOldBurnerStethBalance = _lido.balanceOf(address(_oldBurner));
+
+        _assertProxyImplementation(_locator, _newLocatorImplementation);
+        _assertNewLocatorAddresses();
+        _assertOldBurnerAllowances();
+    }
+
+    function finishUpgrade() external {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        _assertNotExpired();
+        if (_isUpgradeFinished) revert CanOnlyFinishOnce();
+        _isUpgradeFinished = true;
+
+        if (!ALLOW_NON_SINGLE_BLOCK_UPGRADE) {
+            if (_upgradeBlockNumber != block.number) revert StartAndFinishMustBeInSameBlock();
+        }
```

Comment:
:warning:  this hack brakes the contract behavior allowing to finishUpgrade without having it started before